<a href="https://colab.research.google.com/github/unwastefultoday/kpmg-virtual-internship/blob/main/KPMG_Module_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving KPMG_VI_New_raw_data_update_final.xlsx to KPMG_VI_New_raw_data_update_final (1).xlsx


In [2]:
import numpy as np
import pandas as pd


###IMPORTING DATASET

In [3]:
customer_txn = pd.read_excel("KPMG_VI_New_raw_data_update_final.xlsx", "Transactions")
new_customers = pd.read_excel("KPMG_VI_New_raw_data_update_final.xlsx", "NewCustomerList")
customer_demographic = pd.read_excel("KPMG_VI_New_raw_data_update_final.xlsx", "CustomerDemographic")
customer_address = pd.read_excel("KPMG_VI_New_raw_data_update_final.xlsx", "CustomerAddress")

#SHEET 1

In [4]:
customer_txn.head()
#column headers are not proper

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   ...              Unnamed: 12
0                                     transaction_id                                                                                                                             ...  product_first_sold_date
1                                                  1                                                                                                                             ...                    41245
2                                                  2                                                                                                                             ...                    41701
3                                                  3                                                                                                                             ...                    36361
4                                                  4                                                                                                                             ...                    36145

[5 rows x 13 columns]

In [5]:
customer_txn.columns =customer_txn.iloc[0]
customer_txn = customer_txn[1:]
customer_txn.head()

0 transaction_id product_id  ... standard_cost product_first_sold_date
1              1          2  ...         53.62                   41245
2              2          3  ...        388.92                   41701
3              3         37  ...        248.82                   36361
4              4         88  ...         381.1                   36145
5              5         78  ...        709.48                   42226

[5 rows x 13 columns]

In [6]:
customer_txn.info()
'''
datatypes of various columns need to be fixed.
txn id, product id, custumer_id,to int
 list_price, standard cost as float
txn date, product first sold date as date
'''

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   transaction_id           20000 non-null  object
 1   product_id               20000 non-null  object
 2   customer_id              20000 non-null  object
 3   transaction_date         20000 non-null  object
 4   online_order             19640 non-null  object
 5   order_status             20000 non-null  object
 6   brand                    19803 non-null  object
 7   product_line             19803 non-null  object
 8   product_class            19803 non-null  object
 9   product_size             19803 non-null  object
 10  list_price               20000 non-null  object
 11  standard_cost            19803 non-null  object
 12  product_first_sold_date  19803 non-null  object
dtypes: object(13)
memory usage: 2.0+ MB


'\ndatatypes of various columns need to be fixed.\ntxn id, product id, custumer_id,to int\n list_price, standard cost as float\ntxn date, product first sold date as date\n'

In [7]:
customer_txn['transaction_id'] = customer_txn['transaction_id'].astype(int)
customer_txn['product_id'] = customer_txn['product_id'].astype(int)
customer_txn['customer_id'] = customer_txn['customer_id'].astype(int)
customer_txn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   transaction_id           20000 non-null  int64 
 1   product_id               20000 non-null  int64 
 2   customer_id              20000 non-null  int64 
 3   transaction_date         20000 non-null  object
 4   online_order             19640 non-null  object
 5   order_status             20000 non-null  object
 6   brand                    19803 non-null  object
 7   product_line             19803 non-null  object
 8   product_class            19803 non-null  object
 9   product_size             19803 non-null  object
 10  list_price               20000 non-null  object
 11  standard_cost            19803 non-null  object
 12  product_first_sold_date  19803 non-null  object
dtypes: int64(3), object(10)
memory usage: 2.0+ MB


In [8]:
customer_txn['list_price'] = customer_txn['list_price'].astype(float)
customer_txn['standard_cost'] = customer_txn['standard_cost'].astype(float)
customer_txn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  int64  
 1   product_id               20000 non-null  int64  
 2   customer_id              20000 non-null  int64  
 3   transaction_date         20000 non-null  object 
 4   online_order             19640 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  float64
 11  standard_cost            19803 non-null  float64
 12  product_first_sold_date  19803 non-null  object 
dtypes: float64(2), int64(3), object(8)
memory usage: 2.0+ MB


In [9]:
customer_txn['transaction_date'] =pd.to_datetime(customer_txn['transaction_date'])
customer_txn['product_first_sold_date'] = customer_txn['product_first_sold_date'].astype(float)
customer_txn['product_first_sold_date'] = pd.to_datetime(customer_txn['product_first_sold_date'], unit='s')
customer_txn.head()
#dates aren't correct in product_first_sold_date column, filled with a default value

0  transaction_id  product_id  ...  standard_cost product_first_sold_date
1               1           2  ...          53.62     1970-01-01 11:27:25
2               2           3  ...         388.92     1970-01-01 11:35:01
3               3          37  ...         248.82     1970-01-01 10:06:01
4               4          88  ...         381.10     1970-01-01 10:02:25
5               5          78  ...         709.48     1970-01-01 11:43:46

[5 rows x 13 columns]

###CHECKING FOR MISSING VALUES

In [10]:
#total number of entries
customer_txn.shape

(20000, 13)

In [11]:
#no. of missing entries
customer_txn.isnull().sum()

0
transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [12]:
#since maxmimum no. of missing values in any column is 360/20000=1.8% of all values, we can safely delete all missing values without compromising quality.
customer_txn.dropna(axis=0,inplace=True)
customer_txn.isnull().sum()

0
transaction_id             0
product_id                 0
customer_id                0
transaction_date           0
online_order               0
order_status               0
brand                      0
product_line               0
product_class              0
product_size               0
list_price                 0
standard_cost              0
product_first_sold_date    0
dtype: int64

###checking for redundant categories in categorical variables, if any.
columns-online order, order status , brand, product line, product class, product size

In [13]:
customer_txn['online_order'].value_counts()

True     9739
False    9706
Name: online_order, dtype: int64

In [14]:
customer_txn['order_status'].value_counts()

Approved     19273
Cancelled      172
Name: order_status, dtype: int64

In [15]:
customer_txn['brand'].value_counts()

Solex             4169
WeareA2B          3245
Giant Bicycles    3244
OHM Cycles        2993
Trek Bicycles     2931
Norco Bicycles    2863
Name: brand, dtype: int64

In [16]:
customer_txn['product_line'].value_counts()

Standard    13920
Road         3894
Touring      1213
Mountain      418
Name: product_line, dtype: int64

In [17]:
customer_txn['product_class'].value_counts()

medium    13587
high       2952
low        2906
Name: product_class, dtype: int64

In [18]:
customer_txn['product_size'].value_counts()

medium    12767
large      3900
small      2778
Name: product_size, dtype: int64

no issues found, checking for duplicate values

In [19]:
customer_txn[customer_txn.duplicated()].sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


0
transaction_id    0.0
product_id        0.0
customer_id       0.0
online_order      0.0
order_status      0.0
brand             0.0
product_line      0.0
product_class     0.0
product_size      0.0
list_price        0.0
standard_cost     0.0
dtype: float64

###SHEET 2

In [20]:
new_customers.head()
#same issue as prev table

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   ... Unnamed: 22
0                                         first_name                                                                                                                             ...       Value
1                                            Chickie                                                                                                                             ...     1.71875
2                                              Morly                                                                                                                             ...     1.71875
3                                            Ardelis                                                                                                                             ...     1.71875
4                                             Lucine                                                                                                                             ...    1.703125

[5 rows x 23 columns]

In [21]:
new_customers.columns =new_customers.iloc[0]
new_customers = new_customers[1:]
new_customers.head()

0 first_name  last_name  gender  ...  NaN Rank     Value
1    Chickie    Brister    Male  ...  1.0    1   1.71875
2      Morly     Genery    Male  ...  1.0    1   1.71875
3    Ardelis  Forrester  Female  ...  1.0    1   1.71875
4     Lucine      Stutt  Female  ...  4.0    4  1.703125
5    Melinda     Hadlee  Female  ...  4.0    4  1.703125

[5 rows x 23 columns]

In [22]:
new_customers.info()
'''
5 columns have no  name
dob column has incorrect datatype(datetime)
value has incorrect datatype(float)
'''

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 1 to 1000
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   first_name                           1000 non-null   object 
 1   last_name                            971 non-null    object 
 2   gender                               1000 non-null   object 
 3   past_3_years_bike_related_purchases  1000 non-null   object 
 4   DOB                                  983 non-null    object 
 5   job_title                            894 non-null    object 
 6   job_industry_category                835 non-null    object 
 7   wealth_segment                       1000 non-null   object 
 8   deceased_indicator                   1000 non-null   object 
 9   owns_car                             1000 non-null   object 
 10  tenure                               1000 non-null   object 
 11  address                       

'\n5 columns have no  name\ndob column has incorrect datatype(datetime)\nvalue has incorrect datatype(float)\n'

In [23]:
new_customers['Value'] = new_customers['Value'].astype(float)

new_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 1 to 1000
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   first_name                           1000 non-null   object 
 1   last_name                            971 non-null    object 
 2   gender                               1000 non-null   object 
 3   past_3_years_bike_related_purchases  1000 non-null   object 
 4   DOB                                  983 non-null    object 
 5   job_title                            894 non-null    object 
 6   job_industry_category                835 non-null    object 
 7   wealth_segment                       1000 non-null   object 
 8   deceased_indicator                   1000 non-null   object 
 9   owns_car                             1000 non-null   object 
 10  tenure                               1000 non-null   object 
 11  address                       

In [24]:
new_customers.DOB = pd.to_datetime(new_customers.DOB)
new_customers.head()

0 first_name  last_name  gender  ...  NaN Rank     Value
1    Chickie    Brister    Male  ...  1.0    1  1.718750
2      Morly     Genery    Male  ...  1.0    1  1.718750
3    Ardelis  Forrester  Female  ...  1.0    1  1.718750
4     Lucine      Stutt  Female  ...  4.0    4  1.703125
5    Melinda     Hadlee  Female  ...  4.0    4  1.703125

[5 rows x 23 columns]

###checking for null values

In [25]:
new_customers.shape
#nan values in dob and last name can be deleted since 1. they're less in number, 2. no reliable way to replace missing data
# while those in job title and ibdustry category will need to be imputed since they're over 10% of the data

(1000, 23)

In [26]:
new_customers.isnull().sum()

0
first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
NaN                                      0
NaN                                      0
NaN                                      0
NaN                                      0
NaN                                      0
Rank                                     0
Value                                    0
dtype: in

In [27]:
new_customers.job_title.value_counts(),new_customers.job_industry_category.value_counts()
#since modal values are rather close to the 2nd most occuring values and since missing values are>=10%, imputing by mode may negativelt affect the analysis. 
#Hence what we can do is, we can create a separate category called 'unknown' for these 2 columns.

(Associate Professor         15
 Environmental Tech          14
 Software Consultant         14
 Chief Design Engineer       13
 Assistant Manager           12
                             ..
 Accountant II                1
 Programmer IV                1
 Administrative Officer       1
 Accounting Assistant III     1
 Web Developer I              1
 Name: job_title, Length: 184, dtype: int64, Financial Services    203
 Manufacturing         199
 Health                152
 Retail                 78
 Property               64
 IT                     51
 Entertainment          37
 Argiculture            26
 Telecommunications     25
 Name: job_industry_category, dtype: int64)

###checking for redundant categories
vars- gender , wealth_segment, deceased_indicator,owns_car

In [ ]:
new_customers.gender.value_counts()
new_customers.gender=new_customers.gender.replace("U","Unspecified")

In [ ]:
new_customers.gender.value_counts()

Female         513
Male           470
Unspecified     17
Name: gender, dtype: int64

In [ ]:
new_customers.wealth_segment.value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

In [ ]:
new_customers.deceased_indicator.value_counts()

N    1000
Name: deceased_indicator, dtype: int64

In [ ]:
new_customers.owns_car.value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

In [ ]:
#checking duplicate entries
new_customers[new_customers.duplicated()].sum()

0
first_name                             0.0
last_name                              0.0
gender                                 0.0
past_3_years_bike_related_purchases    0.0
DOB                                    0.0
job_title                              0.0
job_industry_category                  0.0
wealth_segment                         0.0
deceased_indicator                     0.0
owns_car                               0.0
tenure                                 0.0
address                                0.0
postcode                               0.0
state                                  0.0
country                                0.0
property_valuation                     0.0
NaN                                    0.0
NaN                                    0.0
NaN                                    0.0
NaN                                    0.0
NaN                                    0.0
Rank                                   0.0
Value                                  0.0
dtype: fl

###SHEET 3

In [ ]:
customer_demographic.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   ... Unnamed: 12
0                                        customer_id                                                                                                                             ...      tenure
1                                                  1                                                                                                                             ...          11
2                                                  2                                                                                                                             ...          16
3                                                  3                                                                                                                             ...          15
4                                                  4                                                                                                                             ...           7

[5 rows x 13 columns]

In [ ]:
customer_demographic.columns =customer_demographic.iloc[0]
customer_demographic = customer_demographic[1:]
customer_demographic.head()

0 customer_id      first_name  ... owns_car tenure
1           1         Laraine  ...      Yes     11
2           2             Eli  ...      Yes     16
3           3           Arlin  ...      Yes     15
4           4          Talbot  ...       No      7
5           5  Sheila-kathryn  ...      Yes      8

[5 rows x 13 columns]

In [ ]:
customer_demographic.info()
'past 3 years column should be int'
'dob as date'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 1 to 4000
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   customer_id                          4000 non-null   object
 1   first_name                           4000 non-null   object
 2   last_name                            3875 non-null   object
 3   gender                               4000 non-null   object
 4   past_3_years_bike_related_purchases  4000 non-null   object
 5   DOB                                  3913 non-null   object
 6   job_title                            3494 non-null   object
 7   job_industry_category                3344 non-null   object
 8   wealth_segment                       4000 non-null   object
 9   deceased_indicator                   4000 non-null   object
 10  default                              3698 non-null   object
 11  owns_car                             4000 n

In [ ]:
customer_demographic.past_3_years_bike_related_purchases=customer_demographic.past_3_years_bike_related_purchases.astype(int)
customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 1 to 4000
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   customer_id                          4000 non-null   object
 1   first_name                           4000 non-null   object
 2   last_name                            3875 non-null   object
 3   gender                               4000 non-null   object
 4   past_3_years_bike_related_purchases  4000 non-null   int64 
 5   DOB                                  3913 non-null   object
 6   job_title                            3494 non-null   object
 7   job_industry_category                3344 non-null   object
 8   wealth_segment                       4000 non-null   object
 9   deceased_indicator                   4000 non-null   object
 10  default                              3698 non-null   object
 11  owns_car                             4000 n

In [ ]:
customer_demographic.DOB=pd.to_datetime(customer_demographic.DOB)
customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 1 to 4000
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   object        
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [ ]:
customer_demographic.head()

0 customer_id      first_name  ... owns_car tenure
1           1         Laraine  ...      Yes     11
2           2             Eli  ...      Yes     16
3           3           Arlin  ...      Yes     15
4           4          Talbot  ...       No      7
5           5  Sheila-kathryn  ...      Yes      8

[5 rows x 13 columns]

In [ ]:
#missing values
customer_demographic.shape

(4000, 13)

In [ ]:
customer_demographic.isnull().sum()

0
customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [ ]:
###testing categories
###categorical variables-gender,wealth segment,default
customer_demographic.gender.value_counts()

In [ ]:
customer_demographic.gender=customer_demographic.gender.replace("U","Unspecified").replace('F','Female').replace('Femal','Female').replace('M','Male')

In [ ]:
customer_demographic.gender.value_counts()

Female         2039
Male           1873
Unspecified      88
Name: gender, dtype: int64

In [ ]:
customer_demographic.wealth_segment.value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

In [ ]:
customer_demographic.default.value_counts()
#contains a large amount of junk values

100                                       113
1                                         112
-1                                        111
-100                                       99
â°â´âµâââ                                  53
                                         ... 
ð¾ ð ð ð ð ð ð ð                           31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: default, Length: 90, dtype: int64

###analysing 4th sheet

In [ ]:
customer_address.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   ...          Unnamed: 5
0                                        customer_id                                                                                                                             ...  property_valuation
1                                                  1                                                                                                                             ...                  10
2                                                  2                                                                                                                             ...                  10
3                                                  4                                                                                                                             ...                   9
4                                                  5                                                                                                                             ...                   4

[5 rows x 6 columns]

In [ ]:
customer_address.columns =customer_address.iloc[0]
customer_address = customer_address[1:]
customer_address.head()

0 customer_id              address  ...    country property_valuation
1           1   060 Morning Avenue  ...  Australia                 10
2           2  6 Meadow Vale Court  ...  Australia                 10
3           4   0 Holy Cross Court  ...  Australia                  9
4           5  17979 Del Mar Point  ...  Australia                  4
5           6     9 Oakridge Court  ...  Australia                  9

[5 rows x 6 columns]

In [ ]:
customer_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 1 to 3999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   object
 1   address             3999 non-null   object
 2   postcode            3999 non-null   object
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   object
dtypes: object(6)
memory usage: 187.6+ KB


In [ ]:
customer_address.postcode=customer_address.postcode.astype(int)
customer_address.property_valuation = customer_address.property_valuation.astype(int) 
customer_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 1 to 3999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   object
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(2), object(4)
memory usage: 187.6+ KB


In [ ]:
customer_address.shape
#no missing values

(3999, 6)

###FINAL COMMENTS ON DATA QUALITY ANALYSIS, SHEETWISE-
common observation- data types of each column in every dataset is seen to be 'object' due to shifting of title row to the 2nd row. Correction - removal of the first row.
1. customer transaction-

There are missing values in the table, percentage of which ranges from 0% to 1.8%. Since the number of NaN values are less, the rows containing NaN values may be deleted without compromising the data. 

Columns in which missing values were detected-

online_order-360 missing values

brand-197 missing values

product_line-197 missing values

product_class-197 missing values

product_size-197 missing values

standard_cost-197 missing values

product_first_sold_date-197 missing values

2. new customers-

There are missing values in the table, percentage of which ranges from 0% to 16.5%. Here, the rows containing nan values in last_name and DOB columns can be deleted since they contain the least number of missing values, while in the rest of the columns i.e. job_title and job_industry_category missing values should be imputed since the number of nan values there are significantly large(>=10%) to affect the quality of data analysis if discarded.

Columns in which missing values were detected-

last_name-29 missing values

DOB-17 missing values

job_title-106 missing values

job_industry_category-165 missing values

Also, alongwith 'Male' and 'Female', certain fields contained 'U' as an entry in the 'Gender' column. For better quality of analysis, 'U' should be replaced with something specific. 


3. customer demographic-

There are missing values in the table, percentage of which ranges from 0% to 16.4%. The approach to treat missing values here will be same as in the  previous table, with removal of missing data in columns of last_name, DOB, default and tenure, with imputation in the rest.

Columns in which missing values were detected-

last name- 125 missing values

DOB-87 missing values

job_title-506 missing values

job_industry_category-656 missing values

default-302 missing values

tenure-87 missing values

Moreover, the default column contains a lot of unintelligible values which can provide no insights whatsoever. Dropping this column would be the best way forward.

In the end, there were inconsistent levels present in the 'Gender' column resulting from use of acronyms and full forms together, alongwith typos. This can be treated by replacing them with proper values.

No anomalies were noted in the fourth sheet(customer address) as such, except for the common observation mentioned above.




In [ ]:
customer_address[customer_address.duplicated()].sum()

0
customer_id           0.0
address               0.0
postcode              0.0
state                 0.0
country               0.0
property_valuation    0.0
dtype: float64